### PR TITLE
added iterator.seek()

### PR DESCRIPTION
**Pull request by mafintosh**
I need to be able to seek an iterator for something I'm implementing for dat so I've added support for that.
We store keys like this

```
<key>!<change number> = <value>
```

because that allows to checkout previous versions of a key based on a change number (a change number basically represents the state of the database at a current point in time)

Currently we do this like this

``` js
var rs = db.createReadStream({
  gt: key + '!',
  lte: key + '!' + lexint.pack(changeNumber, 'hex'),
  limit:1,
  reverse:true
})

rs.once('data', function (data) {
  // this is the value we want
})
```

But this means that we have to do a read-stream (iterator) for every single lookup. Using `iterator.seek(key)` we can just have a single iterator and do it like this

``` js
var ite = db.iterator({reverse:true})
// ...
ite.seek(key + '!' + lexint.pack(changeNumber, 'hex'))
ite.next(console.log)
// ...
ite.seek(anotherKey + '!' + lexint.pack(changeNumber, 'hex'))
ite.next(console.log)
```

### CODE REVIEW

**Comment by mcollina** [comment]
I'm +1, but you should add a couple of unit tests.

I'm curious, do you need this to seek backward? What's the specific use
case?

I'm not for having this in AbstractLevelDown, because on some backends it
might be hard to implement.
Il giorno ven 29 mag 2015 alle 19:29 Mathias Buus <notifications@github.com>
ha scritto:

**Comment by mafintosh** [comment]
@mcollina i'll add some tests. i've added our use-case to above thread. we basically use "find nearest" a lot in dat since that allows us to implement fast rollbacks of database to previous state (assuming the database is build on top of an append-only log of some kind)

**Comment by mcollina** [comment]
Are you planning to support multiple down implementations? My only question
is if we need this on all implementations.
Il giorno ven 29 mag 2015 alle 19:54 Mathias Buus <notifications@github.com>
ha scritto:

**Comment by juliangruber** [comment]
+1 for adding it here, we can later decide if we need it in other backends

**Comment by ralphtheninja** [comment]
This is a great addition. I'm +1

**Comment by juliangruber** [comment]
the last commit shows that we should add a couple tests for this^^

**Comment by mafintosh** [comment]
@juliangruber yes :) i'm on it - planning on finishing this up today

**Comment by mafintosh** [comment]
Alright. Added a couple of tests and rebased the commits down to a single one. @mcollina @juliangruber @ralphtheninja thoughts?

**Comment by mcollina** [comment]
LGTM.

Why that this.cache = null?
Il giorno dom 31 mag 2015 alle 13:04 Mathias Buus <notifications@github.com>
ha scritto:

**Comment by mafintosh** [comment]
Actually hold up on merging. There might be an edge case when seeking near the end of a database. I'll add a test for that as well

**Comment by ralphtheninja** [comment]
I'll look more into it tonight, but a few quick things:

* it should be documented
* we should validate the key argument in `iterator.js`, what happens if `key` undefined etc? (+ tests for that)
* it's a bit unclear to me what's happening on the c++ side when using the `seeking` boolean, it feels a bit hacky (but is most likely also related to me not knowing how we have implemented it)

**Comment by mafintosh** [comment]
Alright, I've edge case when iterating near the end of the database so this is LGTM from my end.

@ralphtheninja i've added some docs and it now throws if the key is not a string. the seeking boolean is there since the `Read` method was expecting the iterator to only be seeked when it was created and this was the easiest way to communicate to that method that the iterator was reseeked. There is already a `nexting` flag in there so I thought this was fine. there is also a test for an invalid argument passed to seek as well.

@mcollina we need to clear the cache since that might contain wrong values after you seek

File: README.md [hunk 1]
```diff
@@ -230,6 +231,14 @@ Otherwise, the `callback` function will be called with the following 3 arguments
 
 
 --------------------------------------------------------
+<a name="iterator_seek"></a>
+### iterator#seek(key)
+<code>seek()</code> is an instance method on an existing iterator object, used to seek the underlying LevelDB iterator to a given key.
+
+by calling <code>seek(key)</code> subsequent calls to <code>next(cb)</code> will return key/values larger or smaller than <code>key</code>
```
Comment (juliangruber) [inline README.md]: I suggest updating this to

By calling <code>seek(key)</code>, subsequent calls to <code>next(cb)</code> will return key/values larger or smaller than <code>key</code>, based on your <code>reverse</code> setting in the iterator constructor.

Comment (juliangruber) [inline README.md]: other than that, LVGTM!

**Comment by mafintosh** [comment]
@ralphtheninja @juliangruber i've added your changes

**Comment by juliangruber** [comment]
+1

**Comment by ralphtheninja** [comment]
lgtm!

@mafintosh Merge when you feel like it :)

**Comment by mafintosh** [comment]
Merged. Thanks everyone!

**Comment by ralphtheninja** [comment]
```
$ npm publish
+ leveldown@1.2.0
```

**Comment by mafintosh** [comment]
In case anyone is interested in my use-case for adding seek, checkout https://github.com/mafintosh/rollbackdb
